### PR TITLE
Support type narrowing with casted parameters

### PR DESCRIPTION
**Pull request by staabm**
closes https://github.com/phpstan/phpstan/issues/8714

https://3v4l.org/LPe72

### CODE REVIEW

File: src/Type/Php/CtypeDigitFunctionTypeSpecifyingExtension.php [hunk 1]
```diff
@@ -54,7 +56,18 @@ public function specifyTypes(FunctionReflection $functionReflection, FuncCall $n
 			]);
 		}
 
-		return $this->typeSpecifier->create($node->getArgs()[0]->value, TypeCombinator::union(...$types), $context, false, $scope);
+		$unionType = TypeCombinator::union(...$types);
+		$specifiedTypes = $this->typeSpecifier->create($exprArg, $unionType, $context, false, $scope);
+
+		if ($context->true() && $exprArg instanceof Cast) {
```
Comment (staabm) [inline src/Type/Php/CtypeDigitFunctionTypeSpecifyingExtension.php:62]: my initial attempt was to do this directly within the `TypeSpecifier` ... but I then got the feeling that the type narrowing of a cast-expression is not a concept which can be applied to all specifying extensions.

Comment (ondrejmirtes) [reply src/Type/Php/CtypeDigitFunctionTypeSpecifyingExtension.php:62]: Yeah, well, `is_string((string) $int)` is always true, but it doesn't make `$int` a string :D

File: src/Type/Php/CtypeDigitFunctionTypeSpecifyingExtension.php [hunk 2]
```diff
@@ -54,7 +56,18 @@ public function specifyTypes(FunctionReflection $functionReflection, FuncCall $n
...
+			$castOriginType = TypeCombinator::intersect($scope->getType($exprArg->expr), $unionType);
+
+			$specifiedTypes = $specifiedTypes->unionWith(
+				$this->typeSpecifier->create($exprArg->expr, $castOriginType, $context, false, $scope)
```
Comment (ondrejmirtes) [inline src/Type/Php/CtypeDigitFunctionTypeSpecifyingExtension.php]: Are you sure you can't just use `$unionType` instead of `$castOriginType` here in the 2nd arg?

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]
Also please test the negative context.

File: src/Type/Php/CtypeDigitFunctionTypeSpecifyingExtension.php (hunk 2 above)
Comment (staabm) [reply src/Type/Php/CtypeDigitFunctionTypeSpecifyingExtension.php]: good point. fixed.

**Comment by ondrejmirtes** [comment]
Thank you.

**Comment by ondrejmirtes** [comment]
Thank you.
